### ISSUE TITLE
DFAHook for the last layer

### ISSUE THREAD

**Issue by yhy258** [Florian-BACHO/FDFA#2]
Hello, Thank you for sharing your code.

According to the shared model codes, it seems that DFAHook is not applied to the last layer of the models.
In this case, isn't the last layer (classifier head) trained with BP?

**Comment by Florian-BACHO** [comment]
Hi,

The DFA hooks are used to bypass the automatic differentiation of pytorch and perform the linear projection of output errors to the hidden layer.

However, the last layer (i.e. the output layer) doesn't require any feedback projection as the errors are already available from the loss function. Hence, we do not use any DFA hook for the last layer. However, we use the output hook that allows us to catch the errors at the outputs for the linear projections to the hidden ones.

I hope this is clear enough. If not, feel free to ask any further questions.

**Comment by yhy258** [comment]
Thank you for your kind reply.

In my opinion, one of the most important properties of DFA is "asymmetry" learning.
But, in this case, doesn't the last layer learn in symmetric way rather than using random fixed matrix?

Additionally, I experienced the significant performance decline when applying the DFA hook to the last layer..
I would like to ask your opinion about this phenomenon.

**Comment by yhy258** [comment]
I have a follow-up question about the last layer’s gradients in the Hook setup.
Let me provide an example at the last layer.

W3h2 + b3 = y_p
loss function : L(y_p, y) 

And according to your code utilizing hook modules, the forward process is like this:
... h2 -> Hook(h2, B) -> W3h2 + b3 (Last layer)-> L
To my best knowledge, the gradient output of the last layer would be [W3^T e].
In this scenario, the transpose weight information might be reflected when applying hook modules (B[W3^T e], something like this..)

So, If the above example is true, I guess It is hard to say it is a genuine DFA.
Are there any errors in my example?

**Comment by Florian-BACHO** [comment]
Hi !

> the transpose weight information might be reflected when applying hook modules (B[W3^T e], something like this..)

No. The output layer has a dedicated hook called `OutputTrainingHook`. Its job is to set the placeholder` grad_at_output` tensor sent to all the `DFATrainingHook` objects. Essentially, every network implemented have a `grad_at_output` tensor created before the forward pass:
```python
def forward(self, x):
        grad_at_output = torch.zeros([x.shape[0], 10], requires_grad=False).to(self.device)
```

This tensor is sent to every hook during inference:
```python
x = fc_dfa(x, dir_der_at_output, grad_at_output)
```

The output hook saves this tensor during the forward pass and set it with the gradient tensor received from the loss function during the backward pass:
```python
class OutputHookFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, dir_der_at_output, grad_at_output, train_mode):
        ctx.grad_out = grad_at_output  # Save the gradient placeholder tensor
        ...

    @staticmethod
    def backward(ctx, grad_output):
        grad_at_output = ctx.grad_out  # Get the gradient placeholder tensor

        grad_at_output[:grad_output.shape[0], :].data.copy_(grad_output.data)  # Copy the data from the gradient received from the loss

        return grad_output, None, None, None  # Forward the gradient to the last layer
```

Finally, each DFA hook also receive the `grad_at_output` placeholder tensor during the forward pass, and uses it to project the errors with the feedback matrices:
```python
class DFAHookFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, feedbacks, dir_der_at_output, grad_at_output, train_mode):
        ...
        ctx.grad_at_output = grad_at_output # Save the gradient placeholder tensor
        ...

    @staticmethod
    def backward(ctx, grad_output):
        # Get the gradient placeholder tensor
        grad_at_output = ctx.grad_at_output
        grad_at_output = grad_at_output[:grad_output.shape[0], :]

        # Project output errors (the grad_at_output tensor)
        feedbacks_view = feedbacks.view(-1, np.prod(feedbacks.shape[1:]))
        grad_output_est = grad_at_output.mm(feedbacks_view).view(grad_output.shape)
        return grad_output_est, None, None, None, None  # Forward the projected errors to the hidden layer
```

Note that the `backward` method of `DFAHookFunction` does not use the `grad_output` tensor, which corresponds to the errors computed by the backpropagation. Instead it **only** uses the `grad_at_output` tensor, which stores the errors at the outputs, set by the output hook. Therefore, the errors for the hidden layer in your example would be something like Be and not B[W3^T e].

Also, if the algorithm would do  B[W3^T e], it would probably run into dimension mismatches between B and W3^T e because hidden layers do not have the same size as the output layer.

I hope this is clear enough.

**Comment by yhy258** [comment]
Thank you for your sincere reply!
I entirely underatand :)
